src/commons/authentication/mod.rs

```diff
@@ -38,31 +37,110 @@ pub struct AuthenticationClassSpec {
 #[serde(rename_all = "camelCase")]
 #[allow(clippy::large_enum_variant)]
 pub enum AuthenticationClassProvider {
-    Ldap(LdapAuthenticationProvider),
-    Tls(TlsAuthenticationProvider),
-    Static(StaticAuthenticationProvider),
+    Static(static_::AuthenticationProvider),
+    Ldap(ldap::AuthenticationProvider),
+    Oidc(oidc::AuthenticationProvider),
+    Tls(tls::AuthenticationProvider),
 }
 
 impl AuthenticationClass {
     pub async fn resolve(
         client: &Client,
         authentication_class_name: &str,
-    ) -> Result<AuthenticationClass, Error> {
+    ) -> OperatorResult<AuthenticationClass> {
         client
             .get::<AuthenticationClass>(authentication_class_name, &()) // AuthenticationClass has ClusterScope
             .await
     }
 }
 
+/// Common [`ClientAuthenticationDetails`] which is specified at the client/
+/// product cluster level. It provides a name (key) to resolve a particular
+/// [`AuthenticationClass`]. Additionally, it provides authentication provider
+/// specific configuration (OIDC and LDAP for example).
+///
+/// If the product needs additional (product specific) authentication options,
+/// it is recommended to wrap this struct and use `#[serde(flatten)]` on the
+/// field.
+///
+/// Additionally, it might be the case that special fields are needed in the
+/// contained structs, such as [`oidc::ClientAuthenticationOptions`]. To be able
+/// to add custom fields in that structs without serde(flattening) multiple structs,
+/// they are generic, so you can add additional attributes if needed.
+///
+/// ### Example
+///
+/// ```
+/// # use schemars::JsonSchema;
+/// # use serde::{Deserialize, Serialize};
+/// use stackable_operator::commons::authentication::ClientAuthenticationDetails;
+///
+/// #[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
+/// #[serde(rename_all = "camelCase")]
+/// pub struct SupersetAuthenticationClass {
+///     pub user_registration_role: String,
+///     pub user_registration: bool,
+///
+///     #[serde(flatten)]
+///     pub common: ClientAuthenticationDetails,
+/// }
+/// ```
+#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+#[schemars(description = "")]
+pub struct ClientAuthenticationDetails<O = ()> {
+    /// A name/key which references an authentication class. To get the concrete
+    /// [`AuthenticationClass`], we must resolve it. This resolution can be
+    /// achieved by using [`ClientAuthenticationDetails::resolve_class`].
+    #[serde(rename = "authenticationClass")]
+    authentication_class_ref: String,
+
+    /// This field contains authentication provider specific configuration. It
+    /// is flattened into the final CRD.
```

Comment:
It is not flattened anymore, so we should remove that from the comment.

Reply:
Good catch, was a leftover. Fixed in 3bad60c